### PR TITLE
feat(eth-client): Generalize RPC client

### PR DESCRIPTION
**Pull request by slowli**
## What ❔

- Generalizes RPC client so that the same client is used both for L1 and L2.
- Splits client metrics by the network accordingly.
- Makes RPC traits network-safe by introducing a network associated with a client.

## Why ❔

Improves current and future code reuse.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `zk spellcheck`.

### CODE REVIEW

File: core/lib/web3_decl/src/client/mock.rs [hunk 1]
```diff
@@ -1,73 +1,202 @@
...
+        let req: Req = serde_json::from_value(req)?;
+        let resp = self.inner.call(req).await?;
+        Ok(serde_json::to_value(resp)?)
+    }
+}
+
+/// Request handler for [`MockClient`]. Implemented automatically for sync and async closures
+/// taking zero to 3 de-serializable args (you might want to specify their types explicitly)
+/// and returning `Result<Resp, Error>`, where `Resp` is serializable.
+pub trait Handler<Req, const ASYNC: bool>: Send + Sync + 'static {
```
Comment (slowli) [inline core/lib/web3_decl/src/client/mock.rs:70]: Maybe a bit of an overkill, but it allows to define mocks really easily and fluently. (I may introduce a similar approach for mocking contracts.) Can extract in a separate PR if it makes sense. Inspired [by `axum`](https://docs.rs/axum/latest/axum/handler/trait.Handler.html) (I think other HTTP frameworks use similar handler extraction as well).

Comment (popzxc) [reply core/lib/web3_decl/src/client/mock.rs:70]: I am (as always) a bit grumpy when it comes to increasing cognitive complexity, but this change does seem to be useful. Given that it's implemented already, let's keep it here.

File: core/lib/web3_decl/src/client/mod.rs [hunk 2]
```diff
@@ -107,37 +110,54 @@ pub trait TaggedClient {
 //   [a boxed cloneable version of services](https://docs.rs/tower/latest/tower/util/struct.BoxCloneService.html),
 //   but it doesn't fit (it's unconditionally `!Sync`, and `Sync` is required for `HttpClient<_>` to implement RPC traits).
 #[derive(Clone)]
-pub struct L2Client<C = HttpClient> {
+pub struct Client<Net, C = HttpClient> {
```
[the remark this answers is not in the document]
Comment (popzxc) [reply core/lib/web3_decl/src/client/mod.rs:113]: Fine by me. We can always rename if anyone feels confused (AFAIU we don't have plans to promise any kind of code API backward compatibility even after reaching crates.io).

File: core/lib/web3_decl/src/client/shared.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+use async_trait::async_trait;
+use jsonrpsee::core::{
+    client::{BatchResponse, ClientT, Error, Subscription, SubscriptionClientT},
+    params::BatchRequestBuilder,
+    traits::ToRpcParams,
+};
+use serde::de::DeserializeOwned;
+
+#[derive(Debug)]
+pub struct Shared<T>(Arc<T>);
```
Comment (popzxc) [inline core/lib/web3_decl/src/client/shared.rs:14]: Do we need it because of orphan rule?

File: core/lib/web3_decl/src/namespaces/eth.rs [hunk 4]
```diff
@@ -1,25 +1,26 @@
-use jsonrpsee::{
-    core::{RpcResult, SubscriptionResult},
-    proc_macros::rpc,
-};
+#[cfg_attr(not(feature = "server"), allow(unused_imports))]
```
[the remark this answers is not in the document]
Comment (popzxc) [reply core/lib/web3_decl/src/namespaces/eth.rs:1]: Prolly we can ship a fix upstream :)

**Review by popzxc — APPROVED** [review]

File: core/lib/web3_decl/src/client/shared.rs (hunk 3 above)
Comment (slowli) [reply core/lib/web3_decl/src/client/shared.rs:14]: Yep, both `Arc` and the traits we want to implement are defined externally.

**Review by perekopskiy — APPROVED** [review]

File: core/lib/web3_decl/Cargo.toml [hunk 5]
```diff
@@ -34,6 +35,6 @@ test-casing.workspace = true
 tokio = { workspace = true, features = ["rt", "test-util"] }
 
 [features]
-default = ["server", "client"]
+default = []
```
[the remark this answers is not in the document]
Comment (popzxc) [reply core/lib/web3_decl/Cargo.toml:38]: I'm fine with removing features.

[the remark this answers is not in the document]
Comment (slowli) [reply core/lib/web3_decl/Cargo.toml:38]: Will do in a separate PR.
